src/geometry/manifold/bump_function.lean

```diff
@@ -0,0 +1,599 @@
+def choice [t2_space M] [sigma_compact_space M] :
+  smooth_bump_covering I (univ : set M) :=
+choice_set I univ is_closed_univ
+
+variables {I M}
+
+protected lemma locally_finite : locally_finite (λ i, support (f i)) := f.locally_finite'
+
+protected lemma point_finite (x : M) : {i | f i x ≠ 0}.finite :=
+f.locally_finite.point_finite x
+
+lemma mem_chart_at_source_of_eq_one {i : f.ι} {x : M} (h : f i x = 1) :
+  x ∈ (chart_at H (f.c i)).source :=
+(f i).support_subset_source $ by simp [h]
+
+lemma mem_ext_chart_at_source_of_eq_one {i : f.ι} {x : M} (h : f i x = 1) :
+  x ∈ (ext_chart_at I (f.c i)).source :=
+by { rw ext_chart_at_source, exact f.mem_chart_at_source_of_eq_one h }
+
+/-- Index of a bump function such that `f i =ᶠ[𝓝 x] 1`. -/
+def ind (x : M) (hx : x ∈ s) : f.ι := (f.eventually_eq_one' x hx).some
+
+lemma eventually_eq_one (x : M) (hx : x ∈ s) : f (f.ind x hx) =ᶠ[𝓝 x] 1 :=
+(f.eventually_eq_one' x hx).some_spec
+
+lemma apply_ind (x : M) (hx : x ∈ s) : f (f.ind x hx) x = 1 :=
+(f.eventually_eq_one x hx).eq_of_nhds
+
+lemma mem_support_ind (x : M) (hx : x ∈ s) : x ∈ support (f $ f.ind x hx) :=
+by simp [f.apply_ind x hx]
+
+lemma mem_chart_at_ind_source (x : M) (hx : x ∈ s) :
+  x ∈ (chart_at H (f.c (f.ind x hx))).source :=
+f.mem_chart_at_source_of_eq_one (f.apply_ind x hx)
+
+lemma mem_ext_chart_at_ind_source (x : M) (hx : x ∈ s) :
+  x ∈ (ext_chart_at I (f.c (f.ind x hx))).source :=
+f.mem_ext_chart_at_source_of_eq_one (f.apply_ind x hx)
+
+section embedding
+
+/-!
+### Whitney embedding theorem
+
+In this section we prove a version of the Whitney embedding theorem: for any compact real manifold
+`M`, for sufficiently large `n` there exists a smooth embedding `M → ℝ^n`.
+-/
+
+instance fintype_ι_of_compact [compact_space M] : fintype f.ι :=
+f.locally_finite.fintype_of_compact $ λ i, (f i).nonempty_support
+
+variables [t2_space M] [fintype f.ι]
+
+/-- Smooth embedding of `M` into `(E × ℝ) ^ f.ι`. -/
+def embedding_pi_tangent : C^∞⟮I, M; 𝓘(ℝ, f.ι → (E × ℝ)), f.ι → (E × ℝ)⟯ :=
+{ to_fun := λ x i, (f i x • ext_chart_at I (f.c i) x, f i x),
+  times_cont_mdiff_to_fun := times_cont_mdiff_pi_space.2 $ λ i,
+    ((f i).smooth_smul times_cont_mdiff_on_ext_chart_at).prod_mk_space ((f i).smooth) }
+
+local attribute [simp] lemma embedding_pi_tangent_coe :
+  ⇑f.embedding_pi_tangent = λ x i, (f i x • ext_chart_at I (f.c i) x, f i x) :=
+rfl
+
+lemma embedding_pi_tangent_inj_on : inj_on f.embedding_pi_tangent s :=
+begin
+  intros x hx y hy h,
+  simp only [embedding_pi_tangent_coe, funext_iff] at h,
+  obtain ⟨h₁, h₂⟩ := prod.mk.inj_iff.1 (h (f.ind x hx)),
+  rw [f.apply_ind x hx] at h₂,
+  rw [← h₂, f.apply_ind x hx, one_smul, one_smul] at h₁,
+  have := f.mem_ext_chart_at_source_of_eq_one h₂.symm,
+  exact (ext_chart_at I (f.c _)).inj_on (f.mem_ext_chart_at_ind_source x hx) this h₁
+end
+
+lemma embedding_pi_tangent_injective (f : smooth_bump_covering I (univ : set M))
+  [fintype f.ι] :
```

Comment:
The previous lemmas deal with `smooth_bump_covering I s` (so that you can apply it to a compact subset in a non-compact manifold). This lemma is about `smooth_bump_covering I univ`. I added `inj_on` lemma for the same reason.

Reply:
Ah, ok. Could you use `fs : smooth_bump_covering I s` and `f : smooth_bump_covering I (univ : set M)` to help your reader when he is not careful enough, just like I was?